feature: springdoc integration

klopfdreh opened the pull request:
This is a PR to implement springdoc. It relates to https://github.com/spring-cloud/spring-cloud-dataflow/issues/4820.

I was not able to build the project completely at my local machine, maybe someone can help me with this. I get some errors related to sun jdk which makes sense, because I use OpenJDK.

The default behavior is that springdoc is deactivated by default. I had to implement a EnvironmentPostProcessor to achieve this.  Only if you set the corresponding features to the value "true", the API and the UI will be available. See: https://springdoc.org/properties.html (springdoc.swagger-ui.enabled / springdoc.api-docs.enabled)

Note: I implemented a filter and didn't use a configurer in the mvc configuration, because this will lead to errors when opening task logs, because it is not wrapped into quotes.

I implemented it for 2.9.x but it can easily be cherrypicked to main.

I made the api-docs public without any login required.

Licenses:
Swagger: Version 2.0, January 2004
Springdoc: Version 2.0, January 2004

klopfdreh commented:
@ilayaperumalg build passed - the PR can be reviewed, now. 😁

onobc reviewed (changes requested):
Thanks for the contribution @klopfdreh !

I have left some comments inline. Here are a couple of points of concern as well:

### Testing
It is ok that there are no tests in it for this 1st round as we may morph direction a bit. However, we will want to have the following tests in before this gets merged:

- Unit tests where applicable 
- IT for auto-config disabled by default
- IT for auto-config enabled by setting enabled props
- IT for any other OpenAPI configs that we add (customizers, etc...)
- I am sure there are a few more we will want to add when we get to this stage

### Custom Configuration
We will probably want to make sure we allow users to set [properties](https://springdoc.org/properties.html) to customize OpenAPI. We should make sure that we respect these properties. I would expect this to just work OOTB but would be great thing to either add an IT for that does some lightweight customization or at very least a manual test. 

### Doc updates 
We should add some high level docs about this. 
- How to enable / disable
- How to customize via openapi props (point to their doc)
- How to view / use


### Security
My biggest concern is what we want to do for security OOTB. OpenAPI is integrated in very well w/ Spring Boot. This is awesome and opens up some cool features. However, it brings to mind a spidey cliche quote "With great power comes great responsibility". We have to watch out what we allow OpenAPI to expose. One example is the ability to show the actuator endpoints via setting `springdoc.show-actuator`.  Any other built-in features that we should be aware of? 

We have an upcoming feature that leverages secured Actuator endpoints and we need to be sure that whatever we do here does not allow the user to subvert the security required by that feature. @dturanski your thoughts here would be greatly appreciated.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowSwaggerApiEnvironmentPostProcessor.java:
```diff
@@ -0,0 +1,66 @@
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.PropertiesPropertySource;
+
+public class DataFlowSwaggerApiEnvironmentPostProcessor implements EnvironmentPostProcessor {
```
Great suggestion! This would be indeed a better solution. 👍

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowSwaggerApiEnvironmentPostProcessor.java:
```diff
@@ -0,0 +1,66 @@
+
+public class DataFlowSwaggerApiEnvironmentPostProcessor implements EnvironmentPostProcessor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DataFlowSwaggerApiEnvironmentPostProcessor.class);
```
I also think to use default props approach would be better than this.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/DataFlowSwaggerApiDocsJsonDecodeFilter.java:
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2016-2018 the original author or authors.
```
Yes good point.

klopfdreh replied:
I changed it to 2022

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/DataFlowSwaggerApiDocsJsonDecodeFilter.java:
```diff
@@ -0,0 +1,82 @@
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+import org.springframework.web.util.ContentCachingResponseWrapper;
+
```
This filter is required because there is an issue similar to this one: https://github.com/springdoc/springdoc-openapi/issues/624. The api-docs endpoints returns escaped JSON. Because Spring Cloud Data Flow relies on escaped JSON content for example to show the output logs of a task execution I implemented this filter.

onobc replied:
Thanks for the details @klopfdreh . We can basically copy/paste your comment as the javadoc - well stated. 

> The endpoints springdoc.swagger-ui.configUrl and springdoc.api-docs.path return escaped JSON.

Is this the default behavior for Springdoc or is this because of some config settings in SCDF to satisfy the "Spring Cloud Data Flow relies on escaped JSON content" requirement?

klopfdreh replied:
Normally the endpoint of api-docs.path returns a valid JSON which is the default behavior. I assume that because SCDF communicates through JSON with the backend and requires it to be escaped to parse it in Angular this behavior is requires for SCDF. This is the issue here and the reason why the filter is required

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/DataFlowSwaggerApiDocsJsonDecodeFilter.java:
```diff
@@ -0,0 +1,82 @@
+
+@Component
+@Order
+public class DataFlowSwaggerApiDocsJsonDecodeFilter implements Filter {
```
See: https://github.com/spring-cloud/spring-cloud-dataflow/pull/4836#discussion_r817376744

klopfdreh replied:
I had a look at the wrapper and they only store the original request / response but don't change anything to it. in the else block the original request and response is used for chaining. Because as for the if statements the default values of the properties are used I am sure we can consider it to be bullet-proof. wdyt?

klopfdreh replied:
This has been solved by using the FilterRegistrationBean.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataflowOAuthSecurityConfiguration.java:
```diff
@@ -32,9 +36,42 @@
 
+	private static final String SWAGGER_UI_CONTEXT = "/swagger-ui/**";
+
+	@Value("${springdoc.api-docs.path:/v3/api-docs}")
```
I think the security concerns are very appropriate about this feature. Because we needed the API to be available public I implemented it that way, but the changes could also be easily removed. I guess the server needs to be requested with a bearer token / with a previous login if security is enabled in this case.

If this would be a better way to provide the feature for the majority of SCDF users I would be fine 👍  

I thinks `springdoc.show-actuator` is only be considered when the actuator endpoints are activated via configuration in the application.yaml

onobc replied:
> I just checked if it is possible to remove the security configuration. 

Did you move the code to a WebSecurityCustomizer? I did not mean to get rid of it entirely, just move it out of here.

onobc replied:
> I thinks springdoc.show-actuator is only be considered when the actuator endpoints are activated via configuration in the application.yaml

I too think we are safe here but just wanted to call it out as an area we would need to verify closely.

onobc replied:
> Because we needed the API to be available public I implemented it that way, 

Yep, got it and I understand why you did it. I would have done the same thing. And this may be the way we end up going. Lets see what the rest of the team chimes in on etc. For now, lets assume we will leave it open and just move this code into a separate customizer. Wdyt?

klopfdreh replied:
> Did you move the code to a WebSecurityCustomizer? I did not mean to get rid of it entirely, just move it out of here.
I tested it locally so no changes to this branch were made. 😁 

I moved the code to a WebSecurityCustomizer and restored `DataflowOAuthSecurityConfiguration` to the original implementation.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/DataFlowSwaggerApiDocsJsonDecodeFilter.java:
```diff
@@ -0,0 +1,82 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(DataFlowSwaggerApiDocsJsonDecodeFilter.class);
+
+    @Value("${springdoc.api-docs.path:/v3/api-docs}")
```
Done

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataflowOAuthSecurityConfiguration.java:
```diff
@@ -32,9 +36,42 @@
+
+	@Override
+	public void configure(WebSecurity web) {
+		String apiDocsPathContext = StringUtils.substringBeforeLast(apiDocsPath, "/");
```
I refactored the code to use vanilla java with no additional classes.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/DataFlowSwaggerApiDocsJsonDecodeFilter.java:
```diff
@@ -0,0 +1,82 @@
+import org.springframework.web.util.ContentCachingResponseWrapper;
+
+@Component
+@Order
```
I removed the Order annotation because the default behavior is ok.

klopfdreh commented:
I renamed the classes into SpringDoc*, because it includes OpenApi and Swagger (the default integration) and I added a Unit-Test for the Filter in which you can see the behavior. 

One little question to ensure this: Properties of ConfigurationProperties can be written in camel case and snake case right? So the configuration of `swaggerUi` can be written as `swagger-ui` or `swaggerUi` - I just want to ensure this.

@onobc can you try the branch and see if SpringDoc is working for you?

onobc commented:
@klopfdreh thanks for all of the updates. I will try to get back to this today (worst case w/in 24hrs).

onobc commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocWebSecurityCustomizer.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @author Tobias Soloschenko
+ */
+@Configuration
```
I think we may want to name this more generic such as `SpringDocConfiguration` as we will want to put the FilterRegistrationBean in it as well (probably). Instead of `implements WebSecurityCustomizer` we can create a bean method that returns a lambda for the customizer such as:
```java
    @Bean
    public WebSecurityCustomizer springDocCustomizer() { 
        return (webSecurity -> {
            ....
        });
    }
```

klopfdreh replied:
Done

onobc reviewed (commented):
@klopfdreh looking good. I did not get to the test but wanted to give you some feedback as I may not be able to get back to this for 1-2 days. 

Thanks

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilter.java:
```diff
@@ -0,0 +1,92 @@
+            // if api-docs path is requested, use wrapper classes, so that the body gets cached.
+            chain.doFilter(httpServletRequestWrapper, httpServletResponseWrapper);
+
+            ServletOutputStream outputStream = httpServletResponseWrapper.getResponse().getOutputStream();
```
You don't need to close the ServletOutputStream, because this is handled by the server.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfigurationProperties.java:
```diff
@@ -0,0 +1,127 @@
+ */
+@Configuration
+@ConfigurationProperties(prefix = "springdoc")
+public class SpringDocConfigurationProperties {
```
Ah that is what you mean. I am going to use the Properties of SpringDoc. 👍

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilter.java:
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2016-2018 the original author or authors.
```
Done

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocWebSecurityCustomizer.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public void customize(WebSecurity webSecurity) {
+        String apiDocsPath = springDocConfigurationProperties.getApiDocs().getPath();
+        String apiDocsPathContext = apiDocsPath.substring(0, apiDocsPath.lastIndexOf("/"));
```
I just want everything under /v3/ and /swagger-ui/ included into the exception for security. I put everything into the config, now.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilter.java:
```diff
@@ -0,0 +1,92 @@
+        String apiDocsPath = springDocConfigurationProperties.getApiDocs().getPath();
+        String swaggerUiConfigUrl = springDocConfigurationProperties.getSwaggerUi().getConfigUrl();
+
+        String apiDocsPathContext = apiDocsPath.substring(0, apiDocsPath.lastIndexOf("/"));
```
Done - much better this way.

klopfdreh commented:
The build issue is because a SNAPSHOT dependency could not be resolved. I did not changed anything to the dependencies.

onobc commented:
> One little question to ensure this: Properties of ConfigurationProperties can be written in camel case and snake case right? So the configuration of `swaggerUi` can be written as `swagger-ui` or `swaggerUi` - I just want to ensure this.

Absolutely. What you are referring to is one of the beautiful features of config props, [relaxed binding](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.external-config.typesafe-configuration-properties.relaxed-binding). Kebab case is the preferred choice when in yaml/properties. I believe you meant that by "snake case". Snake case is the python style underscores such as `swagger_ui_` IIRC.

onobc reviewed (commented):
@klopfdreh 
I believe we are very close on this one . Nice work! 

This current round of review is mostly of the nit variety. The one bigger comment is around the auto-config and conditional loading. 

Also, I think we should only put this in main (2.10 release). I was going to bring this up earlier but it we were in middle of back/forth and it would only be distracting at that point. As you alluded to in your comment from the beginning - cherrypicking to main would not be a problem. Lets go ahead and move this to main. 

---- 

I will go back through my [initial list](https://github.com/spring-cloud/spring-cloud-dataflow/pull/4836#pullrequestreview-897055287) of things we will want to add/cover next:

### Testing
- [x] Unit tests where applicable
- [x] IT for auto-config disabled by default
- [ ] IT for auto-config enabled by setting enabled props
- [ ] IT for auto-config one prop enabled, one prop disabled
- [ ] IT for custom configuration using the built-in Springdoc properties path changes

For the auto-config tests, the `ApplicationContextRunner` is a beautiful tool. If you have not seen it, I would be happy to handle the start of these tests and let you flesh the others out. It will give a base of where to start w/ the ApplicationContextRunner. Just lmk. 

### Docs
- [ ] We should add some high level docs about this.
    - How to enable / disable
    - How to customize via openapi props (point to their doc)
    - How to view / use

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    @Bean
+    public WebSecurityCustomizer springDocCustomizer() {
+        return (webSecurity -> {
```
Because only one method is called lambda can be written as

`return (interfaceArg -> method(....));`

no need for curly braces in this case.

onobc replied:
Sure. I meant the parens around the arg. 

```java
(arg) -> foo()
```

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+     *
+     * @return a web security customizer with security settings for SpringDoc
+     */
+    @Bean
```
Good point I added it.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+            String apiDocsPathContext = getApiDocsPathContext();
+            webSecurity.ignoring().antMatchers(
+                    SWAGGER_UI_CONTEXT,
+                    apiDocsPathContext + "/**",
```
Yep - I inlined it.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml:
```diff
@@ -1,3 +1,8 @@
+springdoc:
```
Moved it to the end. 👍

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilter.java:
```diff
@@ -0,0 +1,71 @@
+        final HttpServletRequestWrapper httpServletRequestWrapper = new HttpServletRequestWrapper((HttpServletRequest) request);
+        final ContentCachingResponseWrapper httpServletResponseWrapper = new ContentCachingResponseWrapper((HttpServletResponse) response);
+
+        // if api-docs path is requested, use wrapper classes, so that the body gets cached.
```
Oh forgot this leftover. Thanks!

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilter.java:
```diff
@@ -0,0 +1,71 @@
+        ServletOutputStream outputStream = httpServletResponseWrapper.getResponse().getOutputStream();
+
+        LOG.debug("Request for Swagger api-docs detected - unescaping json content.");
+        String content = new String(httpServletResponseWrapper.getContentAsByteArray(), StandardCharsets.UTF_8);
```
the endpoints will all the time responding with a valid JSON, so we don't need to think about this.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilter.java:
```diff
@@ -0,0 +1,71 @@
+        // Replaces first and last quote
+        content = content.substring(1, content.length() - 1);
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Using decoded JSON for serving api-docs: {}", content);
```
I implemented it that way so that you are able to see the replaced content in scenarios in which there might be an error because of the filter. This way you could put this class in log level trace and see what is going on. This could be done in any server and used to trace down any issues that might occur.

I also used trace, because a lot of devs are using debug on root level of the logging which would let the whole content to be printed out all the time you access the endpoint which could be irritating.

I would stay with it.

onobc replied:
Absolutely agree w/ TRACE. I was talking about printing the payload to the log. I don't think its much of an issue in this case but was bringing it up.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Creates a filter registration bean which decodes the JSON of ApiDocs and SwaggerUi so that the SpringDoc frontend is able
```
I changed it a bit by mentioning the `SpringDocJsonDecodeFilter`, but would stay with a bit more text. This text describes why the url patterns are applied.

klopfdreh commented on spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilterTest.java:
```diff
@@ -0,0 +1,63 @@
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+
+import org.junit.Before;
```
no problem - upgraded it. 👍

klopfdreh commented on spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilterTest.java:
```diff
@@ -0,0 +1,63 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class SpringDocJsonDecodeFilterTest {
```
done

klopfdreh commented on spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilterTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    public void doFilterTest() throws ServletException, IOException {
+        MockHttpServletResponse mockHttpServletResponse = new MockHttpServletResponse();
```
done

klopfdreh commented on spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilterTest.java:
```diff
@@ -0,0 +1,63 @@
+                super.doFilter(request, response);
+            }
+        };
+        springDocJsonDecodeFilter.doFilter(mockHttpServletRequest, mockHttpServletResponse, mockFilterChain);
```
👍

klopfdreh commented on spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilterTest.java:
```diff
@@ -0,0 +1,63 @@
+            }
+        };
+        springDocJsonDecodeFilter.doFilter(mockHttpServletRequest, mockHttpServletResponse, mockFilterChain);
+        assertEquals(OPENAPI_JSON_UNESCAPED_CONTENT, mockHttpServletResponse.getContentAsString());
```
I thought to stick with something which comes from junit would not require more dependencies, but anyway I used AssertJ now. 👍

onobc replied:
SpringBoot test starters brings it in so it can be pretty much thought of as included across the board. You will find it used in almost all Spring projects.

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+ * @author Tobias Soloschenko
+ */
+@Configuration(proxyBeanMethods = false)
+public class SpringDocConfiguration {
```
Done

klopfdreh commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+     * @return a web security customizer with security settings for SpringDoc
+     */
+    @Bean
+    public WebSecurityCustomizer springDocCustomizer() {
```
I added qualifier to the two beans in the configuration

klopfdreh commented:
The tests and the docs have to wait until I have more time at weekend - have to do my daily work now. 👍

onobc commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocAutoConfiguration.java:
```diff
@@ -58,30 +61,29 @@ public SpringDocConfiguration(SpringDocConfigProperties springDocConfigPropertie
      * @return a web security customizer with security settings for SpringDoc
      */
     @Bean
+    @Qualifier("springDocWebSecurityCustomizer")
```
I would prefer just naming the bean this. The name serves as a qualifier in the abscence of a qualifier. If something were actually using the qualifier then I would go that route.

klopfdreh replied:
I used `@Bean("springDocWebSecurityCustomizer")`

onobc commented on spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/SpringDocAutoConfiguration.java:
```diff
@@ -58,30 +61,29 @@ public SpringDocConfiguration(SpringDocConfigProperties springDocConfigPropertie
      */
     @Bean
-    public FilterRegistrationBean<SpringDocJsonDecodeFilter> jsonDecodeFilterRegistration() {
+    @Qualifier("springDocJsonDecodeFilterRegistration")
```
We can remove `@Qualifier` here as the name is the same so this is redundant.

klopfdreh replied:
Okay - I just removed it. 👍

onobc commented on spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/support/SpringDocJsonDecodeFilterTest.java:
```diff
@@ -56,8 +52,8 @@ public void doFilter(ServletRequest request, ServletResponse response) throws IO
-        springDocJsonDecodeFilter.doFilter(mockHttpServletRequest, mockHttpServletResponse, mockFilterChain);
-        assertEquals(OPENAPI_JSON_UNESCAPED_CONTENT, mockHttpServletResponse.getContentAsString());
+        new SpringDocJsonDecodeFilter().doFilter(mockHttpServletRequest, mockHttpServletResponse, mockFilterChain);
+        assertThat(OPENAPI_JSON_UNESCAPED_CONTENT).isEqualTo(mockHttpServletResponse.getContentAsString());
```
**[TinyNit]** the assertion target should be the `actual` rather than the `expected`.

klopfdreh replied:
done

onobc commented:
> The tests and the docs have to wait until I have more time at weekend - have to do my daily work now. 👍

Sounds good @klopfdreh . I will also pull down and take a test drive and also try to push the start of the auto-config tests up before you get back to it.

onobc commented:
Hi @klopfdreh, I have not forgotten about my [next steps](https://github.com/spring-cloud/spring-cloud-dataflow/pull/4836#issuecomment-1060752288) and should get to it sometime today (week got busy).

klopfdreh commented:
Hey @onobc no hurry 😃

onobc commented:
Hi @klopfdreh 
So I pulled down and started writing the auto-config test and noticed something strange. 

There is something going on w/ the Springdoc config props - as in they are coming to us w/o any defaults in them. The docs suggest [various defaults](https://springdoc.org/properties.html) but these defaults are not in the actual config props but rather in scattered locations that uses the config props (afaict). 

This results in our customizer adding a bunch of null patterns as well as NPE in our filter registration as it relies on config-url. 

Did you encounter this? 

One workaround is to set the advertised defaults in our defaults file as such:
```

springdoc:
  api-docs:
    enabled: false
    path: /v3/api-docs
  swagger-ui:
    enabled: false
    path: /swagger-ui.html
    config-url: /v3/api-docs/swagger-config
    validator-url: https://validator.swagger.io/validator
    oauth2-redirect-url: /swagger-ui/oauth2-redirect.html
```

However, I would like to be able to explain whats going on before we go that route.

klopfdreh commented:
Hey @onobc - previously I set the default values accordingly to the documentation of springdoc. I would suggest to do the same with the settings we are going to use for our purpose.

https://springdoc.org/properties.html

One big issue I have: I use an Apple Silicon and was not able to build / run the server completely.

The @Value implementation I was able to test in another scenario which worked. You can see it in the commit history, but in this case I applied the default values.

I am not totally sure how springdoc initializes the default values but I found all Constants in this class here: https://github.com/springdoc/springdoc-openapi/blob/master/springdoc-openapi-common/src/main/java/org/springdoc/core/Constants.java

onobc commented:
Closed in favor of https://github.com/spring-cloud/spring-cloud-dataflow/pull/4855
